src/ethereum_watcher/ethereum_utils.rs

```diff
@@ -0,0 +1,138 @@
+use ethers::providers::{Provider, Http, Middleware};
+use ethers::prelude::k256::ecdsa::SigningKey;
+use std::sync::Arc;
+use anyhow::{Result, anyhow};
+use std::convert::TryFrom;
+use std::ops::Mul;
+use ethers::abi::AbiEncode;
+use fuels::tx::Bytes32;
+use ethers::prelude::{GasEscalatorMiddleware, Signer, Wallet, Log};
+use ethers::middleware::gas_escalator::{Frequency, GeometricGasPrice};
+use ethers::types::U256;
+
+pub async fn setup_ethereum_provider(
+    ethereum_rpc: &str,
+) -> Result<Arc<GasEscalatorMiddleware<Provider<Http>>>> {
+    // Geometrically increase gas price:
+    // Start with `initial_price`, then increase it every 'every_secs' seconds by a fixed
+    // coefficient. Coefficient defaults to 1.125 (12.5%), the minimum increase for Parity to
+    // replace a transaction. Coefficient can be adjusted, and there is an optional upper limit.
+    let coefficient: f64 = 1.125;
+    let every_secs: u64 = 60;
+    let max_price: Option<i32> = None;
+
+    let geometric_escalator = GeometricGasPrice::new(
+        coefficient,
+        every_secs,
+        max_price,
+    );
+
+    let provider = Provider::<Http>::try_from(ethereum_rpc)?;
+    let provider = GasEscalatorMiddleware::new(
+        provider,
+        geometric_escalator,
+        Frequency::PerBlock,
+    );
+
+    let provider_result = provider.get_chainid().await;
+    match provider_result {
+        Ok(_) => Ok(Arc::new(provider)),
+        Err(e) => Err(anyhow!("Failed to get chain ID: {e}")),
+    }
+}
+
+pub fn setup_ethereum_wallet(
+    ethereum_wallet_key: Option<String>,
+    chain_id: u64,
+) -> Result<(Wallet<SigningKey>, bool)> {
+
+    let mut read_only: bool = false;
+    let key_str = ethereum_wallet_key.unwrap_or_else(|| {
+        read_only = true;
+        "ac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80".to_string()
```

Comment:
Whats this constant?

Reply:
Private key from the fuel-bridge testnet, their code to make the wallet not break

Reply:
placeholder